dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/EventServiceContextBuilder.java

```diff
@@ -0,0 +1,125 @@
+import org.hisp.dhis.trackedentity.TrackedEntityProgramOwnerIds;
+import org.hisp.dhis.trackedentity.TrackedEntityProgramOwnerService;
+import org.hisp.dhis.user.User;
+import org.springframework.stereotype.Service;
+
+@Service
+public class EventServiceContextBuilder
+{
+
+    private final TrackedEntityInstanceService entityInstanceService;
+
+    private final ProgramService programService;
+
+    private final TrackedEntityProgramOwnerService trackedEntityProgramOwnerService;
+
+    private final OrganisationUnitService organisationUnitService;
+
+    public EventServiceContextBuilder( TrackedEntityInstanceService entityInstanceService,
+        ProgramService programService, TrackedEntityProgramOwnerService trackedEntityProgramOwnerService,
+        OrganisationUnitService organisationUnitService )
+    {
+        this.entityInstanceService = entityInstanceService;
+        this.programService = programService;
+        this.trackedEntityProgramOwnerService = trackedEntityProgramOwnerService;
+        this.organisationUnitService = organisationUnitService;
+    }
+
+    public EventContext build( List<EventRow> eventRowList, User user )
+    {
+        Map<String, List<EventRow>> eventsByProgramUid = eventRowList.stream()
+            .collect( Collectors.groupingBy( EventRow::getProgram ) );
+
+        List<String> trackedEntityInstanceUids = eventRowList.stream()
+            .map( EventRow::getTrackedEntityInstance )
+            .filter( Objects::nonNull )
+            .distinct()
+            .collect( Collectors.toList() );
+
+        Map<String, TrackedEntityInstance> trackedEntityInstanceByUid = entityInstanceService
+            .getTrackedEntityInstancesByUid( trackedEntityInstanceUids, user )
+            .stream()
+            .collect( Collectors.toMap(
+                BaseIdentifiableObject::getUid,
+                trackedEntityInstance -> trackedEntityInstance ) );
+
+        Map<String, Program> programsByUid = programService.getPrograms(
+            eventRowList.stream()
+                .map( EventRow::getProgram )
+                .collect( Collectors.toSet() ) )
+            .stream()
+            .collect( Collectors.toMap(
+                BaseIdentifiableObject::getUid,
+                program -> program ) );
+
+        Map<Pair<String, String>, String> orgUnitByTeiUidAndProgramUidPairs = eventsByProgramUid.keySet().stream()
+            .flatMap( programUid -> trackedEntityProgramOwnerService.getTrackedEntityProgramOwnersUidsUsingId(
+                eventsByProgramUid.get( programUid ).stream()
+                    .map( EventRow::getTrackedEntityInstance )
+                    .filter( Objects::nonNull )
+                    .map( trackedEntityInstanceByUid::get )
+                    .map( BaseIdentifiableObject::getId )
+                    .distinct()
+                    .collect( Collectors.toList() ),
+                programsByUid.get( programUid ) ).stream() )
```

Comment:
@larshelge @bobjolliffe I agree with you this part of code is a bit tricky to read, but I had in mind to finish it so I didn't pur much care in readability, I'm sorry. Here we are invoking getTrackedEntityProgramOwners... with a collection of tracked entities uids and a program. This will **not** retrieve all orgUnits in a program